### PR TITLE
Fix script fetching and local path handling in `inspect_dataset` and `inspect_metric`

### LINKED ISSUE #4348: `inspect` functions can't fetch dataset script from the Hub
The `inspect_dataset` and `inspect_metric` functions are unable to retrieve a dataset or metric script from the Hub and store it locally at the specified `local_path`:

```py
>>> from datasets import inspect_dataset
>>> inspect_dataset('rotten_tomatoes', local_path='path/to/my/local/folder')

FileNotFoundError: Couldn't find a dataset script at /content/rotten_tomatoes/rotten_tomatoes.py or any data file in the same directory.
```

### PR DESCRIPTION
**Pull request by mariosasko**
Fix #4348

### CODE REVIEW

**Review by lhoestq — COMMENTED** [review]
Thanks ! Just one comment:

File: src/datasets/download/streaming_download_manager.py [hunk 1]
```diff
@@ -759,7 +759,7 @@ def _iter_from_urlpaths(
                 yield urlpath
             else:
                 for dirpath, dirnames, filenames in xwalk(urlpath, use_auth_token=use_auth_token):
-                    for i, dirname in enumerate(dirnames):
+                    for i, dirname in enumerate(dirnames[:]):
```
Comment (lhoestq) [inline src/datasets/download/streaming_download_manager.py:762]: Why is this needed ? Is it related to the `del dirnames[i]` call ? If so maybe it's simpler to just remove `del dirnames[i]` instead. I feel like this optimization adds more confusion that it brings value

File: src/datasets/download/download_manager.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def _iter_from_paths(cls, urlpaths: Union[str, List[str]]) -> Generator[str, Non
                 yield urlpath
             else:
                 for dirpath, dirnames, filenames in os.walk(urlpath):
-                    for i, dirname in enumerate(dirnames):
+                    for i, dirname in enumerate(dirnames[:]):
```
Comment (lhoestq) [inline src/datasets/download/download_manager.py:131]: same here

File: src/datasets/inspect.py [hunk 3]
```diff
@@ -118,13 +123,24 @@ def inspect_dataset(path: str, local_path: str, download_config: Optional[Downlo
...
+    dataset_module = dataset_module_factory(path, download_config=download_config, **download_kwargs)
+    builder_cls = import_main_class(dataset_module.module_path, dataset=True)
+    module_source_path = inspect.getsourcefile(builder_cls)
+    module_source_dirpath = os.path.dirname(module_source_path)
+    for dirpath, dirnames, filenames in os.walk(module_source_dirpath):
+        dst_dirpath = os.path.join(local_path, os.path.relpath(dirpath, module_source_dirpath))
+        os.makedirs(dst_dirpath, exist_ok=True)
+        for i, dirname in enumerate(dirnames[:]):
+            if dirname == "__pycache__":
+                del dirnames[i]
```
Comment (lhoestq) [inline src/datasets/inspect.py:135]: Same here, and I guess you can ignore directories that start with "." and "__" to be more general, as we do in the download manager
```suggestion
        for i, dirname in enumerate(dirnames):
            if dirname.startswith((".", "__")):
                continue
```

File: src/datasets/inspect.py [hunk 4]
```diff
@@ -143,13 +159,24 @@ def inspect_metric(path: str, local_path: str, download_config: Optional[Downloa
...
+    metric_module = metric_module_factory(path, download_config=download_config, **download_kwargs)
+    builder_cls = import_main_class(metric_module.module_path, dataset=False)
+    module_source_path = inspect.getsourcefile(builder_cls)
+    module_source_dirpath = os.path.dirname(module_source_path)
+    for dirpath, dirnames, filenames in os.walk(module_source_dirpath):
+        dst_dirpath = os.path.join(local_path, os.path.relpath(dirpath, module_source_dirpath))
+        os.makedirs(dst_dirpath, exist_ok=True)
+        for i, dirname in enumerate(dirnames[:]):
+            if dirname == "__pycache__":
+                del dirnames[i]
```
Comment (lhoestq) [inline src/datasets/inspect.py:171]: Same here
```suggestion
        for i, dirname in enumerate(dirnames):
            if dirname.startswith((".", "__")):
                continue
```

**Review by lhoestq — COMMENTED** [review]
Cool thanks, just some nits regarding `dirname[:]` because it can seem confusing why it's needed

File: src/datasets/download/streaming_download_manager.py (hunk 1 above)
Comment (mariosasko) [reply src/datasets/download/streaming_download_manager.py]: Yes. There is a comment that explains this inside the loop... It's outside of it now. And I removed the `[:]` syntax.

**Review by lhoestq — APPROVED** [review]
Thanks !

**Review by lhoestq — CHANGES_REQUESTED** [review]
oh actually it looks like tests failed on windows/py3.7 - removing my approval

**Review by lhoestq — APPROVED** [review]
alright all good now ! thanks

**Comment by mariosasko** [comment]
Added back the `[:]` and a comment to explain why this is needed.
